### ISSUE TITLE
"zlib.h: No such file or directory" but zlib is installed and confirmed via Spack. Help? 

### ISSUE THREAD

**Issue by j-ogas** [spack/spack#7665]
When attempting to install the `serf` package via Spack, it fails saying that zlib.h cannot be found. However, when I type `spack find zlib` I see it is indeed installed. Furthermore, at the beginning of the serf package install it says `zlib is already installed in /spack/opt/spack/linux-debian9-x86_64/gcc-6.3.0/zlib-1.2.11-lgevn3ckxbm4sxwjdoq3k5qfgmpsjt7o`.

I apologize if I have missed something trivial but I have been unable to move past this issue. Please let me know if there is something I did wrong!

### Expected Result

`spack install serf`
installs without error

### Actual Result

```$ /spack/share/spack$ spack install serf
==> apr is already installed in /spack/opt/spack/linux-debian9-x86_64/gcc-6.3.0/apr-1.6.2-uwes5lcvwrh42wapuia5wsx3qaj57z7m
==> libbsd is already installed in /spack/opt/spack/linux-debian9-x86_64/gcc-6.3.0/libbsd-0.8.6-kbcxkq2bbidrapfst4wpcnsictq2irx4
==> expat is already installed in /spack/opt/spack/linux-debian9-x86_64/gcc-6.3.0/expat-2.2.2-64q5ajsmgvlfjkuajo444hcddunymgpo
==> libiconv is already installed in /spack/opt/spack/linux-debian9-x86_64/gcc-6.3.0/libiconv-1.15-3slxs7noyeqoifg574ewtwfkgwo64bse
==> **zlib is already installed in /spack/opt/spack/linux-debian9-x86_64/gcc-6.3.0/zlib-1.2.11-lgevn3ckxbm4sxwjdoq3k5qfgmpsjt7o**
==> openssl is already installed in /spack/opt/spack/linux-debian9-x86_64/gcc-6.3.0/openssl-1.0.2n-qq7fsrjt276e77ml6v3kj7nvz7v4ag5v
==> apr-util is already installed in /spack/opt/spack/linux-debian9-x86_64/gcc-6.3.0/apr-util-1.6.0-e6fghlzggrpkg3esillio4vo2lw46f3c
==> bzip2 is already installed in /spack/opt/spack/linux-debian9-x86_64/gcc-6.3.0/bzip2-1.0.6-pkzuknqz57o4x4riltavgvi56e4bkkvs
==> pkgconf is already installed in /spack/opt/spack/linux-debian9-x86_64/gcc-6.3.0/pkgconf-1.4.0-5elovbvlolf2qem6dyvmxvvkom23unkb
==> ncurses is already installed in /spack/opt/spack/linux-debian9-x86_64/gcc-6.3.0/ncurses-6.0-l6eyrwifxfv3akoimxvhkq7rdfoganol
==> readline is already installed in /spack/opt/spack/linux-debian9-x86_64/gcc-6.3.0/readline-7.0-qxxrlvhtch3ch5z5pvqvk6zgnz44zuaa
==> gdbm is already installed in /spack/opt/spack/linux-debian9-x86_64/gcc-6.3.0/gdbm-1.14.1-ghicvnjqeyecj5pfobqw5ohn7dn5wemo
==> sqlite is already installed in /spack/opt/spack/linux-debian9-x86_64/gcc-6.3.0/sqlite-3.22.0-ytrwvm2e6wssqy2mpopkuipewg7lxjnf
==> python is already installed in /spack/opt/spack/linux-debian9-x86_64/gcc-6.3.0/python-2.7.14-wqv2wxmakqkscbp22ylhgdufiilnay3d
==> scons is already installed in /spack/opt/spack/linux-debian9-x86_64/gcc-6.3.0/scons-3.0.1-wld5wum4o3moc6nyx5hjwlohlrbse5bk
==> Installing serf
==> Using cached archive: /spack/var/spack/cache/serf/serf-1.3.9.tar.bz2
==> Staging archive: /spack/var/spack/stage/serf-1.3.9-34o6k34f3eyy4jq4p7vg3yr6kjp2gsxc/serf-1.3.9.tar.bz2
==> Created stage in /spack/var/spack/stage/serf-1.3.9-34o6k34f3eyy4jq4p7vg3yr6kjp2gsxc
==> No patches needed for serf
==> Building serf [SConsPackage]
==> Executing phase: 'build'
==> Error: ProcessError: Command exited with status 2:
    'scons' 'PREFIX=/spack/opt/spack/linux-debian9-x86_64/gcc-6.3.0/serf-1.3.9-34o6k34f3eyy4jq4p7vg3yr6kjp2gsxc' 'APR=/spack/opt/spack/linux-debian9-x86_64/gcc-6.3.0/apr-1.6.2-uwes5lcvwrh42wapuia5wsx3qaj57z7m' 'APU=/spack/opt/spack/linux-debian9-x86_64/gcc-6.3.0/apr-util-1.6.0-e6fghlzggrpkg3esillio4vo2lw46f3c' 'OPENSSL=/spack/opt/spack/linux-debian9-x86_64/gcc-6.3.0/openssl-1.0.2n-qq7fsrjt276e77ml6v3kj7nvz7v4ag5v' 'ZLIB=/spack/opt/spack/linux-debian9-x86_64/gcc-6.3.0/zlib-1.2.11-lgevn3ckxbm4sxwjdoq3k5qfgmpsjt7o' 'DEBUG=no' 'CC=/usr/bin/gcc-6'

1 error found in build log:
     13    /usr/bin/gcc-6 -o buckets/barrier_buckets.o -c -std=c89 -Wdeclaration-after-statement -Wmissing-prototypes -Wall -g -O2 -pthrea
           d -DNDEBUG -DLINUX -D_REENTRANT -D_GNU_SOURCE -I. -I/spack/opt/spack/linux-debian9-x86_64/gcc-6.3.0/apr-1.6.2-uwes5lcvwrh42wapu
           ia5wsx3qaj57z7m/include/apr-1 -I/spack/opt/spack/linux-debian9-x86_64/gcc-6.3.0/apr-util-1.6.0-e6fghlzggrpkg3esillio4vo2lw46f3c
           /include/apr-1 -I/spack/opt/spack/linux-debian9-x86_64/gcc-6.3.0/openssl-1.0.2n-qq7fsrjt276e77ml6v3kj7nvz7v4ag5v/include -I/spa
           ck/opt/spack/linux-debian9-x86_64/gcc-6.3.0/expat-2.2.2-64q5ajsmgvlfjkuajo444hcddunymgpo/include -I/spack/opt/spack/linux-debia
           n9-x86_64/gcc-6.3.0/libiconv-1.15-3slxs7noyeqoifg574ewtwfkgwo64bse/include -I/spack/opt/spack/linux-debian9-x86_64/gcc-6.3.0/op
           enssl-1.0.2n-qq7fsrjt276e77ml6v3kj7nvz7v4ag5v/include buckets/barrier_buckets.c
     14    /usr/bin/gcc-6 -o buckets/buckets.o -c -std=c89 -Wdeclaration-after-statement -Wmissing-prototypes -Wall -g -O2 -pthread -DNDEB
           UG -DLINUX -D_REENTRANT -D_GNU_SOURCE -I. -I/spack/opt/spack/linux-debian9-x86_64/gcc-6.3.0/apr-1.6.2-uwes5lcvwrh42wapuia5wsx3q
           aj57z7m/include/apr-1 -I/spack/opt/spack/linux-debian9-x86_64/gcc-6.3.0/apr-util-1.6.0-e6fghlzggrpkg3esillio4vo2lw46f3c/include
           /apr-1 -I/spack/opt/spack/linux-debian9-x86_64/gcc-6.3.0/openssl-1.0.2n-qq7fsrjt276e77ml6v3kj7nvz7v4ag5v/include -I/spack/opt/s
           pack/linux-debian9-x86_64/gcc-6.3.0/expat-2.2.2-64q5ajsmgvlfjkuajo444hcddunymgpo/include -I/spack/opt/spack/linux-debian9-x86_6
           4/gcc-6.3.0/libiconv-1.15-3slxs7noyeqoifg574ewtwfkgwo64bse/include -I/spack/opt/spack/linux-debian9-x86_64/gcc-6.3.0/openssl-1.
           0.2n-qq7fsrjt276e77ml6v3kj7nvz7v4ag5v/include buckets/buckets.c
     15    /usr/bin/gcc-6 -o buckets/bwtp_buckets.o -c -std=c89 -Wdeclaration-after-statement -Wmissing-prototypes -Wall -g -O2 -pthread -
           DNDEBUG -DLINUX -D_REENTRANT -D_GNU_SOURCE -I. -I/spack/opt/spack/linux-debian9-x86_64/gcc-6.3.0/apr-1.6.2-uwes5lcvwrh42wapuia5
           wsx3qaj57z7m/include/apr-1 -I/spack/opt/spack/linux-debian9-x86_64/gcc-6.3.0/apr-util-1.6.0-e6fghlzggrpkg3esillio4vo2lw46f3c/in
           clude/apr-1 -I/spack/opt/spack/linux-debian9-x86_64/gcc-6.3.0/openssl-1.0.2n-qq7fsrjt276e77ml6v3kj7nvz7v4ag5v/include -I/spack/
           opt/spack/linux-debian9-x86_64/gcc-6.3.0/expat-2.2.2-64q5ajsmgvlfjkuajo444hcddunymgpo/include -I/spack/opt/spack/linux-debian9-
           x86_64/gcc-6.3.0/libiconv-1.15-3slxs7noyeqoifg574ewtwfkgwo64bse/include -I/spack/opt/spack/linux-debian9-x86_64/gcc-6.3.0/opens
           sl-1.0.2n-qq7fsrjt276e77ml6v3kj7nvz7v4ag5v/include buckets/bwtp_buckets.c
     16    /usr/bin/gcc-6 -o buckets/chunk_buckets.o -c -std=c89 -Wdeclaration-after-statement -Wmissing-prototypes -Wall -g -O2 -pthread 
           -DNDEBUG -DLINUX -D_REENTRANT -D_GNU_SOURCE -I. -I/spack/opt/spack/linux-debian9-x86_64/gcc-6.3.0/apr-1.6.2-uwes5lcvwrh42wapuia
           5wsx3qaj57z7m/include/apr-1 -I/spack/opt/spack/linux-debian9-x86_64/gcc-6.3.0/apr-util-1.6.0-e6fghlzggrpkg3esillio4vo2lw46f3c/i
           nclude/apr-1 -I/spack/opt/spack/linux-debian9-x86_64/gcc-6.3.0/openssl-1.0.2n-qq7fsrjt276e77ml6v3kj7nvz7v4ag5v/include -I/spack
           /opt/spack/linux-debian9-x86_64/gcc-6.3.0/expat-2.2.2-64q5ajsmgvlfjkuajo444hcddunymgpo/include -I/spack/opt/spack/linux-debian9
           -x86_64/gcc-6.3.0/libiconv-1.15-3slxs7noyeqoifg574ewtwfkgwo64bse/include -I/spack/opt/spack/linux-debian9-x86_64/gcc-6.3.0/open
           ssl-1.0.2n-qq7fsrjt276e77ml6v3kj7nvz7v4ag5v/include buckets/chunk_buckets.c
     17    /usr/bin/gcc-6 -o buckets/dechunk_buckets.o -c -std=c89 -Wdeclaration-after-statement -Wmissing-prototypes -Wall -g -O2 -pthrea
           d -DNDEBUG -DLINUX -D_REENTRANT -D_GNU_SOURCE -I. -I/spack/opt/spack/linux-debian9-x86_64/gcc-6.3.0/apr-1.6.2-uwes5lcvwrh42wapu
           ia5wsx3qaj57z7m/include/apr-1 -I/spack/opt/spack/linux-debian9-x86_64/gcc-6.3.0/apr-util-1.6.0-e6fghlzggrpkg3esillio4vo2lw46f3c
           /include/apr-1 -I/spack/opt/spack/linux-debian9-x86_64/gcc-6.3.0/openssl-1.0.2n-qq7fsrjt276e77ml6v3kj7nvz7v4ag5v/include -I/spa
           ck/opt/spack/linux-debian9-x86_64/gcc-6.3.0/expat-2.2.2-64q5ajsmgvlfjkuajo444hcddunymgpo/include -I/spack/opt/spack/linux-debia
           n9-x86_64/gcc-6.3.0/libiconv-1.15-3slxs7noyeqoifg574ewtwfkgwo64bse/include -I/spack/opt/spack/linux-debian9-x86_64/gcc-6.3.0/op
           enssl-1.0.2n-qq7fsrjt276e77ml6v3kj7nvz7v4ag5v/include buckets/dechunk_buckets.c
     18    /usr/bin/gcc-6 -o buckets/deflate_buckets.o -c -std=c89 -Wdeclaration-after-statement -Wmissing-prototypes -Wall -g -O2 -pthrea
           d -DNDEBUG -DLINUX -D_REENTRANT -D_GNU_SOURCE -I. -I/spack/opt/spack/linux-debian9-x86_64/gcc-6.3.0/apr-1.6.2-uwes5lcvwrh42wapu
           ia5wsx3qaj57z7m/include/apr-1 -I/spack/opt/spack/linux-debian9-x86_64/gcc-6.3.0/apr-util-1.6.0-e6fghlzggrpkg3esillio4vo2lw46f3c
           /include/apr-1 -I/spack/opt/spack/linux-debian9-x86_64/gcc-6.3.0/openssl-1.0.2n-qq7fsrjt276e77ml6v3kj7nvz7v4ag5v/include -I/spa
           ck/opt/spack/linux-debian9-x86_64/gcc-6.3.0/expat-2.2.2-64q5ajsmgvlfjkuajo444hcddunymgpo/include -I/spack/opt/spack/linux-debia
           n9-x86_64/gcc-6.3.0/libiconv-1.15-3slxs7noyeqoifg574ewtwfkgwo64bse/include -I/spack/opt/spack/linux-debian9-x86_64/gcc-6.3.0/op
           enssl-1.0.2n-qq7fsrjt276e77ml6v3kj7nvz7v4ag5v/include buckets/deflate_buckets.c
  >> 19    buckets/deflate_buckets.c:23:18: fatal error: zlib.h: No such file or directory
     20     #include <zlib.h>
     21                      ^
     22    compilation terminated.
     23    scons: *** [buckets/deflate_buckets.o] Error 1
     24    scons: building terminated because of errors.
```

### Steps to reproduce the issue

```console
$ git clone $SPACK_URL
$ export PATH=$PATH:/spack/bin
$ spack install libelf
$ spack install curl
$ spack bootstrap
$ . /spack/share/spack/setup-env.sh
$ spack install serf
```

### Information on your system

 * Debian 9
 * Followed the "Getting Started" tutorial from the Spack documentation (no custom modules or scriptfu).

**Comment by skosukhin** [comment]
As far as I understand, it's a combination of two problems:
1. We can't use Spack's compiler wrappers, which would add the required -I and -L flags (see #4927).
2. Serf's [SConstruct never adds ZLIB paths on non-Windows platforms](https://www.mail-archive.com/dev%40serf.apache.org/msg01359.html).

The second problem is already solved upstream but not for the latest stable version 1.3.9, which means we need to generate a patch file and apply it before the installation.

**Comment by skosukhin** [comment]
Ok, this had been fixed once (#1037) but got broken again later (#4936).

**Comment by adamjstewart** [comment]
My bad for breaking the package, thanks for the fix!

**Comment by adamjstewart** [comment]
Closed by #7667.
